File: SI231b/HM/HM3/code/LS_gradient_decent.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Sun Oct 25 17:47:01 2020

@author: tianyun
"""

import numpy as np
                                              
alpha = 0.01                                  
eps = 1e-6                                    
                                              
x = [1., 2., 3., 4., 5., 6., 7., 8., 9.]      
y = [3., 5., 7., 9., 11., 13., 15., 17., 19.] 

#最普通的loadtxt
A = np.loadtxt('data.txt')
Y = np.loadtxt('lable.txt')

def solve_by_gradient():
    # m is sample num
    m = len(Y)
    a, b= 0, 0
    sse, sse_new = 0, 0
    grad_a, grad_b = 0, 0
    count = 0
    for step in range(100000):
        count += 1
        for i in range(m):
            base = a * x[i] + b - y[i]
            grad_a += x[i] * base
            grad_b += base

            grad_a = grad_a / m
            grad_b = grad_b / m

            a -= alpha * grad_a
            b -= alpha * grad_b

            # loss function: Mean Squared Error, MSE
            # because 2m is a const, so 1/2m can be ignored
            for j in range(m):
                sse_new += (a * x[j] + b - y[j]) ** 2

        if abs(sse_new - sse) < eps:
            break
        else:
            sse = sse_new
    print('{0} * x + {1}'.format(a, b))
    print ("count is: " , count)

solve_by_gradient()

File: SI231b/HM/HM3/code/LS_gradient_decent.m
file1 = 'data.txt';
file2 = 'lable.txt';
delimiterIn = ' ';
headerlinesIn = 0;
A = importdata(file1,delimiterIn,headerlinesIn);
Y = importdata(file2,delimiterIn,headerlinesIn);

alpha = 1e-5;
eps = 1e-2;

data_len = size(Y, 1);
var_len = size(A, 2);

J = 0; J_new = 0;

x = zeros(var_len,1);
grad_x = zeros(var_len,1);

count = 0;

for step = 1:1000000
    count = count + 1;
    grad_x = alpha * A'*(A*x-Y);
    x = x - alpha * grad_x;
    J_new = (A*x-Y)'*(A*x-Y)/2;
    if abs(J_new - J)<eps
        break
    else
        J = J_new;
    end
    
end
